### PR TITLE
Feature/1598 object serialization

### PR DESCRIPTION
**Pull request by davidnich**

### CODE REVIEW

File: doxygen/lang/245_parse_directives.dox.tmpl [hunk 1]
```diff
@@ -1202,6 +1204,23 @@ i+ =4;
...
+    @par Command Line:
+    <tt>-</tt><tt>-pno-reflection</tt>
+
+    @par Parse Option Constant:
+    @ref Qore::PO_NO_REFLECTION
+
+    @par Description:
+    Disallows access to the reflection API provided by the <a href="../../modules/reflection/html/index.html">reflection</a> module.
+
+    @since %Qore
```
Comment (omusil24) [inline doxygen/lang/245_parse_directives.dox.tmpl]: Missing version number

File: lib/Makefile.am [hunk 2]
```diff
@@ -77,6 +77,7 @@ QORE_QPP_TARGETS = QC_Queue.cpp QC_Socket.cpp QC_ReadOnlyFile.cpp QC_File.cpp QC
 	QC_StreamWriter.cpp QC_StreamReader.cpp QC_BufferedStreamReader.cpp \
 	QC_Transform.cpp QC_TransformInputStream.cpp QC_TransformOutputStream.cpp \
 	QC_StdoutOutputStream.cpp QC_StderrOutputStream.cpp \
+    QC_Serializable.cpp \
```
Comment (omusil24) [inline lib/Makefile.am]: TAB should be used in makefile.

File: lib/Makefile.am [hunk 3]
```diff
@@ -329,7 +330,8 @@ libqore_la_SOURCES = \
 	xxhash.cpp \
 	minitest.cpp \
 	FunctionalOperatorInterface.cpp \
-	FunctionalOperator.cpp
+	FunctionalOperator.cpp \
+    QoreSerializable.cpp
```
Comment (omusil24) [inline lib/Makefile.am]: TAB should be used here.

File: lib/QoreSerializable.cpp [hunk 4]
```diff
@@ -0,0 +1,1711 @@
...
+
+    ObjectIndexMap oimap(xsink);
+
+    QoreProgram* pgm = getProgram();
+
+    // load any required modules
+    QoreValue val = h.getKeyValue("_modules");
+    if (val) {
+        assert(val.getType() == NT_LIST);
+        QoreProgram* pgm = getProgram();
```
Comment (omusil24) [inline lib/QoreSerializable.cpp]: Unnecessary second declaration.

File: lib/QoreSerializable.cpp [hunk 5]
```diff
@@ -0,0 +1,1711 @@
...
+                                if (!mem.isTransient()) {
+                                    continue;
+                                }
+                                ValueHolder val(mem.getDefaultValue(xsink), xsink);
+                                if (*xsink) {
+                                    return QoreValue();
+                                }
+                                // skip transient member initialization if there is no expression
+                                if (!val) {
+                                    printd(0, "DESERIALIZE transient member %s::%s has no value\n", mcls.getName(), mi.getName());
```
Comment (omusil24) [inline lib/QoreSerializable.cpp]: Shouldn't this be commented out?

File: lib/QoreSerializable.cpp [hunk 6]
```diff
@@ -0,0 +1,1711 @@
...
+            return writer.writei8((int64_t)i, xsink);
+        }
+    }
+
+    return -1;
+}
+
+int QoreSerializable::serializeBoolToStream(bool b, StreamWriter& writer, ExceptionSink* xsink) {
+    // write data type code to stream
+    if (!writer.writei1(QSSDT_BOOLEAN, xsink)) {
```
Comment (omusil24) [inline lib/QoreSerializable.cpp]: 1-byte per bool could be saved if there was simply constants `QSSDT_BOOLEAN_TRUE` and `QSSDT_BOOLEAN_FALSE`.

Comment (omusil24) [reply lib/QoreSerializable.cpp]: Also the code overhead would be smaller.

File: lib/QoreSerializable.cpp [hunk 7]
```diff
@@ -0,0 +1,1711 @@
...
+    ReferenceHolder<QoreHashNode> h(new QoreHashNode, xsink);
+
+    // read in hash keys
+    for (int64 i = 0; i < size; ++i) {
+        // read key type (must be string)
+        code = reader.readi1(xsink);
+        if (*xsink) {
+            return nullptr;
+        }
+        if (code != QSSDT_UTF8_STRING && code != QSSDT_STRING) {
```
Comment (omusil24) [inline lib/QoreSerializable.cpp]: CODE_IS_STRING macro can be used here.

Comment (sejvlond) [reply lib/QoreSerializable.cpp]: I would vote for as few macros as possible, ideally none :)

**Review by omusil24 — APPROVED** [review]

**Comment by omusil24** [comment]
@sejvlond you can merge now.

File: lib/QoreSerializable.cpp [hunk 8]
```diff
@@ -0,0 +1,1705 @@
...
+#define QSSDT_FLOAT 11
+#define QSSDT_NUMBER 12
+#define QSSDT_BINARY 13
+#define QSSDT_ABSDATE 14
+#define QSSDT_RELDATE 15
+#define QSSDT_NULL 16
+#define QSSDT_NOTHING 17
+
+#define CODE_IS_INT(x) ((x) == QSSDT_INT1 || (x) == QSSDT_INT2 || (x) == QSSDT_INT4 || (x) == QSSDT_INT8)
+#define CODE_IS_STRING(x) ((x) == QSSDT_STRING || (x) == QSSDT_UTF8_STRING)
```
Comment (sejvlond) [inline lib/QoreSerializable.cpp]: is there a reason for defining these constants as defines and macros? how about define them as enums with constexpr inline function?

Comment (davidnich) [reply lib/QoreSerializable.cpp]: done

**Pull request merged**